bd-test-helpers/src/workflow.rs

```diff
@@ -664,3 +704,96 @@ pub fn make_tags(labels: BTreeMap<String, String>) -> LogFields {
+}
+
+pub enum TestFieldType {
+  Single(TestFieldRef),
+  Subtract(TestFieldRef, TestFieldRef),
+  Add(TestFieldRef, TestFieldRef),
+  Multiply(TestFieldRef, TestFieldRef),
+  Divide(TestFieldRef, TestFieldRef),
+}
+
+#[must_use]
+pub fn make_generate_log_action(
+  message: &'static str,
+  fields: &'static [(&'static str, TestFieldType)],
+) -> ActionGenerateLog {
+  fn make_field_ref(test_field_ref: &TestFieldRef) -> ValueReference {
+    ValueReference {
+      value_reference_type: Some(match test_field_ref {
+        TestFieldRef::Fixed(value) => Value_reference_type::Fixed((*value).to_string()),
+        TestFieldRef::FieldFromCurrentLog(field_name) => {
+          Value_reference_type::FieldFromCurrentLog((*field_name).to_string())
+        },
+        TestFieldRef::SavedFieldId(saved_field_id) => {
+          Value_reference_type::SavedFieldId((*saved_field_id).to_string())
+        },
+        TestFieldRef::SavedTimestampId(saved_timestamp_id) => {
+          Value_reference_type::SavedTimestampId((*saved_timestamp_id).to_string())
+        },
+      }),
+      ..Default::default()
+    }
+  }
+
+  ActionGenerateLog {
+    message: message.to_string(),
+    fields: fields
+      .iter()
+      .map(|(name, field_type)| {
+        let generated_field_value_type = match field_type {
+          TestFieldType::Single(field_ref) => {
+            Generated_field_value_type::Single(make_field_ref(field_ref))
+          },
+          TestFieldType::Subtract(lhs, rhs) => {
+            Generated_field_value_type::Subtract(ValueReferencePair {
+              lhs: Some(make_field_ref(lhs)).into(),
+              rhs: Some(make_field_ref(rhs)).into(),
+              ..Default::default()
+            })
+          },
+          TestFieldType::Add(lhs, rhs) => Generated_field_value_type::Add(ValueReferencePair {
+            lhs: Some(make_field_ref(lhs)).into(),
+            rhs: Some(make_field_ref(rhs)).into(),
+            ..Default::default()
+          }),
+          TestFieldType::Multiply(lhs, rhs) => {
+            Generated_field_value_type::Multiply(ValueReferencePair {
+              lhs: Some(make_field_ref(lhs)).into(),
+              rhs: Some(make_field_ref(rhs)).into(),
+              ..Default::default()
+            })
+          },
+          TestFieldType::Divide(lhs, rhs) => {
+            Generated_field_value_type::Divide(ValueReferencePair {
+              lhs: Some(make_field_ref(lhs)).into(),
+              rhs: Some(make_field_ref(rhs)).into(),
+              ..Default::default()
+            })
+          },
+        };
+        GeneratedField {
+          name: (*name).to_string(),
+          generated_field_value_type: Some(generated_field_value_type),
+          ..Default::default()
+        }
+      })
+      .collect(),
+    ..Default::default()
+  }
+}
+
+#[must_use]
+pub fn make_generate_log_action_proto(
+  message: &'static str,
+  fields: &'static [(&'static str, TestFieldType)],
```

Comment:
i would think a bunch of these don't have to specify 'static?

Reply:
It just avoids having to have a lifetime on the enum and then having that go everywhere even if anonymous...